README.md
[![Build Status](https://travis-ci.org/ryanchapman/go-any-proxy.png)](https://travis-ci.org/ryanchapman/go-any-proxy)

# Any Proxy

go-any-proxy is a server that can transparently proxy any tcp connection through an upstream proxy server.  This type
of setup is common in corporate environments.  It is written in golang and has been load tested with 10,000 concurrent
connections successfully on a Vyatta running a 64-bit kernel.

## Travis-CI

Build status can be found at http://travis-ci.org/ryanchapman/go-any-proxy

## More info

For more info, see http://blog.rchapman.org/post/47406142744/transparently-proxying-http-and-https-connections

-Ryan A. Chapman
 Sat Jun  1 19:17:25 MDT 2013


debian/changelog
any-proxy (1.1-MW0FY2) UNRELEASED; urgency=low
  commit ff022c016f09e31671a6d7348dbb1ea621bae8e6
  Author: Ryan A. Chapman <ryan@rchapman.org>
  Date:   Sat Nov 9 13:25:53 2013 -0700
  
      More unit tests for issue #11
  
  commit d5779717053a5c78d8ca4717a825d75f9236642b
  Author: Ryan A. Chapman <ryan@rchapman.org>
  Date:   Sat Nov 9 12:24:05 2013 -0700
  
      Add unit test for issue #11
  
  commit b8a42afd3f7388454f16ee7e0b0277a4a801c5b9
  Merge: f0a85a0 c4a6196
  Author: Ryan A. Chapman <ryan@rchapman.org>
  Date:   Tue Nov 5 15:40:50 2013 -0800
  
      Merge pull request #11 from zdannar/master
      
      Fixed shared memory issue between builds of go.
  
  commit c4a6196e7e4731c9dead820fc49041b9a0935d75
  Author: Zack Dannar <zack.dannar@rightnow.com>
  Date:   Tue Nov 5 16:38:27 2013 -0700
  
      Another debug statement
  
  commit 56ad843a4b53e02982c6e70d6f677d6462e55b6c
  Author: Zack Dannar <zack.dannar@rightnow.com>
  Date:   Tue Nov 5 16:37:25 2013 -0700
  
      Removed debugging statements
  
  commit 6cbaf27efa5e41fdceb598cf0ebd467f767c4108
  Author: Zack Dannar <zack.dannar@rightnow.com>
  Date:   Tue Nov 5 16:34:33 2013 -0700
  
      Fixed directories shared memory issue
  
  commit f0a85a066be44545738500075154a59f9cab6499
  Author: Ryan A. Chapman <ryan@rchapman.org>
  Date:   Sun Nov 3 23:00:55 2013 -0700
  
      Changed net.Dial() to net.DialTCP(). See issue #9 for more details
  
  commit a54c108b26081ae655fbd9bb61bfeab505b95da1
  Author: Ryan A. Chapman <ryan@rchapman.org>
  Date:   Sun Nov 3 04:14:02 2013 -0700
  
      Release version 1.1
  
  commit bdaab100fa7677451165f9301182163d35a866b5
  Author: Ryan A. Chapman <ryan@rchapman.org>
  Date:   Sun Nov 3 04:00:33 2013 -0700
  
      Handle nil net.TCPConn gracefully, meaning without generating a panic
  
  commit 0d613b7d5751240388f02780d922534f7105b5ee
  Author: Ryan A. Chapman <ryan@rchapman.org>
  Date:   Sun Nov 3 03:35:27 2013 -0700
  
      Add unit tests that should catch future issues where
      net.TCPConn.fd is nil (still not sure why it sometimes is nil)
  
  commit 193298758a739fd860950d408f40ec1167d36c41
  Author: Ryan A. Chapman <ryan@rchapman.org>
  Date:   Sun Nov 3 01:55:23 2013 -0700
  
      Bubble up errors in finding original destination ip/port from getOriginalDst to handleConnection(), which is the entry point for the goroutine to handle a connection from a client
  
  commit 2581eb49f7caa8e049047f7f892df4ce5f9558db
  Author: Ryan A. Chapman <ryan@rchapman.org>

version_test.go
package main

import "testing"

func TestVersion(t *testing.T) {
    if BUILDTIMESTAMP <= 1 {
        t.Fatalf("BUILDTIMESTAMP=%v, want > 1", BUILDTIMESTAMP)
    }

    lenBuildUser := len(BUILDUSER)
    if lenBuildUser <= 0 {
        t.Fatalf("len(BUILDUSER)=%v, want > 0; BUILDUSER=%s", lenBuildUser, BUILDUSER)
    }

    lenBuildHost := len(BUILDHOST)
    if lenBuildHost <= 0 {
        t.Fatalf("len(BUILDHOST)=%v, want > 0; BUILDHOST=%s", lenBuildUser, BUILDHOST)
    }
}



make.bash
#!/bin/bash

# Required if you are using a 32-bit vyatta. Otherwise, comment it out
GOOS=linux
#GOARCH=386

function make_version ()
{
    local timestamp=`date +%s`
    local builduser=`id -un`
    local buildhost=`hostname`
cat <<vEOF >$BUILD_DIR/version.go
package main

const BUILDTIMESTAMP = $timestamp
const BUILDUSER      = "$builduser"
const BUILDHOST      = "$buildhost"
vEOF
    echo "Wrote $BUILD_DIR/version.go: timestamp=$timestamp; builduser=$builduser; buildhost=$buildhost"
}

function build ()
{
    make_version
    go build any_proxy.go stats.go version.go
    return $?
}

function build_failed ()
{
    echo "TRAVIS_TEST_RESULT=$TRAVIS_TEST_RESULT"
    echo "Build failed."
    echo "CWD:"
    pwd | sed 's/^/  /g'
    echo "Environment:"
    set | sed 's/^/  /g'
    echo "$BUILD_DIR/version.go:"
    cat $BUILD_DIR/version.go | sed 's/^/  /g'
}

function make_debian_package ()
{
    local VERSION=`./any_proxy -h | head -1 | awk '{print $2 "-" $4}' | tr -d ','`
    echo "any-proxy ($VERSION) UNRELEASED; urgency=low" > debian/changelog
    git log | sed 's/^/  /g' >> debian/changelog
    echo " -- Ryan A. Chapman <ryan@rchapman.org>  `date '+%a, %d %b %Y %H:%M:%S %z'`" >> debian/changelog
    dpkg-buildpackage -d
}

function reindex_debian_packages ()
{
    echo "Rebuilding debian package repos"
    (cd debian/apt && ./reindex_stable.sh)
}

export BUILD_DIR=$2
if [ "$BUILD_DIR" = "" ]; then export BUILD_DIR=.; fi
case $1 in 
  "clean")
    rm -f version.go any_proxy
    ;;
  "version")
    make_version
    ;;
  "build_failed")
    build_failed
    ;;
  "package_write_pubkey")
    TMPF=`mktemp /tmp/anyproxy_pub.XXX`.key
    gpg --export -a "ryan@rchapman.org" > $TMPF
    echo "Wrote pubkey to $TMPF"
    echo "Importing pubkey $TMPF into debian/apt/repo.gpg"
    gpg --no-default-keyring --keyring debian/apt/repo.gpg --import $TMPF
    if [[ $? == 0 ]]; then
        echo "Wrote pubkey successfully to debian/apt/repo.gpg"
    else
        echo "ERROR: could not write pubkey to debian/apt/repo.gpg"
    fi
    rm -f $TMPF
    ;;
  "package")
    if [[ -f any_proxy ]]; then
        echo -n "A built product already exist. Re-packaging existing build (y/n) [y]: "
        read ANS
        if [[ $ANS != n && $ANS != N ]]; then
            make_debian_package
            reindex_debian_packages
        fi
    fi
    ;;
  *)
    build
    ;;
esac



any_proxy.go
//
// any_proxy.go - Transparently proxy a connection using Linux iptables REDIRECT
//
// Copyright (C) 2013 Ryan A. Chapman. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//   1. Redistributions of source code must retain the above copyright notice,
//      this list of conditions and the following disclaimer.
//
//   2. Redistributions in binary form must reproduce the above copyright notice, 
//      this list of conditions and the following disclaimer in the documentation
//      and/or other materials provided with the distribution.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS
// OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//
//
// Tested to 2000 connections/second.  If you turn off logging, you can get 10,000/sec. So logging needs
// to be changed to nonblocking one day.
//
// TODO:
// add num of connected clients to stats
// add ability to print details of each connected client (src,dst,proxy or direct addr) to stats
//
// Ryan A. Chapman, ryan@rchapman.org
// Sun Apr  7 21:04:34 MDT 2013
//


package main

import (
    "bufio"
    "errors"
    "flag"
    "fmt"
    "io"
    log "github.com/ryanchapman/go-any-proxy/flogger"
    "net"
    "os"
    "os/signal"
    "runtime"
    "runtime/pprof"
    "strconv"
    "strings"
    "syscall"
    "time"
)

const VERSION = "1.1"
const SO_ORIGINAL_DST = 80
const DEFAULTLOG = "/var/log/any_proxy.log"
const STATSFILE  = "/var/log/any_proxy.stats"


var gListenAddrPort string
var gProxyServerSpec string
var gDirects string
var gVerbosity int
var gProxyServers []string
var gLogfile string
var gCpuProfile string
var gMemProfile string

type directorFunc func(string) bool
var director func(string) (bool, int)

func init() {
    flag.Usage = func() {
        fmt.Fprintf(os.Stdout, "%s\n\n", versionString())
        fmt.Fprintf(os.Stdout, "usage: %s -l listenaddress -p proxies [-d directs] [-v=N] [-f file] [-c file] [-m file]\n", os.Args[0])
        fmt.Fprintf(os.Stdout, "       Proxies any tcp or udp port transparently using Linux netfilter\n\n")
        fmt.Fprintf(os.Stdout, "Mandatory\n")
        fmt.Fprintf(os.Stdout, "  -l=ADDRPORT      Address and port to listen on (e.g., :3128 or 127.0.0.1:3128)\n")
        fmt.Fprintf(os.Stdout, "Optional\n")
        fmt.Fprintf(os.Stdout, "  -p=PROXIES       Address and ports of upstream proxy servers to use\n")
        fmt.Fprintf(os.Stdout, "                   Multiple address/ports can be specified by separating with commas\n")
        fmt.Fprintf(os.Stdout, "                   (e.g., 10.1.1.1:80,10.2.2.2:3128 would try to proxy requests to a\n")
        fmt.Fprintf(os.Stdout, "                    server listening on port 80 at 10.1.1.1 and if that failed, would\n")
        fmt.Fprintf(os.Stdout, "                    then try port 3128 at 10.2.2.2)\n")
        fmt.Fprintf(os.Stdout, "                   Note that requests are not load balanced. If a request fails to the\n")
        fmt.Fprintf(os.Stdout, "                   first proxy, then the second is tried and so on.\n\n")
        fmt.Fprintf(os.Stdout, "  -d=DIRECTS       List of IP addresses that the proxy should send to directly instead of\n")
        fmt.Fprintf(os.Stdout, "                   to the upstream proxies (e.g., -d 10.1.1.1,10.1.1.2)\n")
        fmt.Fprintf(os.Stdout, "  -v=1             Print debug information to logfile %s\n", DEFAULTLOG)
        fmt.Fprintf(os.Stdout, "  -f=FILE          Log file. If not specified, defaults to %s\n", DEFAULTLOG)
        fmt.Fprintf(os.Stdout, "  -c=FILE          Write a CPU profile to FILE. The pprof program, which is part of Golang's\n")
        fmt.Fprintf(os.Stdout, "                   standard pacakge, can be used to interpret the results. You can invoke pprof\n")
        fmt.Fprintf(os.Stdout, "                   with \"go tool pprof\"\n")
        fmt.Fprintf(os.Stdout, "  -m=FILE          Write a memory profile to FILE. This file can also be interpreted by golang's pprof\n\n")
        fmt.Fprintf(os.Stdout, "any_proxy should be able to achieve 2000 connections/sec with logging on, 10k with logging off (-f=/dev/null).\n")
        fmt.Fprintf(os.Stdout, "Before starting any_proxy, be sure to change the number of available file handles to at least 65535\n")
        fmt.Fprintf(os.Stdout, "with \"ulimit -n 65535\"\n")
        fmt.Fprintf(os.Stdout, "Some other tunables that enable higher performance:\n")
        fmt.Fprintf(os.Stdout, "  net.core.netdev_max_backlog = 2048\n")
        fmt.Fprintf(os.Stdout, "  net.core.somaxconn = 1024\n")
        fmt.Fprintf(os.Stdout, "  net.core.rmem_default = 8388608\n")
        fmt.Fprintf(os.Stdout, "  net.core.rmem_max = 16777216\n")
        fmt.Fprintf(os.Stdout, "  net.core.wmem_max = 16777216\n")
        fmt.Fprintf(os.Stdout, "  net.ipv4.ip_local_port_range = 2000 65000\n")
        fmt.Fprintf(os.Stdout, "  net.ipv4.tcp_window_scaling = 1\n")
        fmt.Fprintf(os.Stdout, "  net.ipv4.tcp_max_syn_backlog = 3240000\n")
        fmt.Fprintf(os.Stdout, "  net.ipv4.tcp_max_tw_buckets = 1440000\n")
        fmt.Fprintf(os.Stdout, "  net.ipv4.tcp_mem = 50576 64768 98152\n")
        fmt.Fprintf(os.Stdout, "  net.ipv4.tcp_rmem = 4096 87380 16777216\n")
        fmt.Fprintf(os.Stdout, "  NOTE: if you see syn flood warnings in your logs, you need to adjust tcp_max_syn_backlog, tcp_synack_retries and tcp_abort_on_overflow\n");
        fmt.Fprintf(os.Stdout, "  net.ipv4.tcp_syncookies = 1\n")
        fmt.Fprintf(os.Stdout, "  net.ipv4.tcp_wmem = 4096 65536 16777216\n")
        fmt.Fprintf(os.Stdout, "  net.ipv4.tcp_congestion_control = cubic\n\n")
        fmt.Fprintf(os.Stdout, "To obtain statistics, send any_proxy signal SIGUSR1. Current stats will be printed to %v\n", STATSFILE)
        fmt.Fprintf(os.Stdout, "Report bugs to <ryan@rchapman.org>.\n") 
    }
    flag.StringVar(&gListenAddrPort,  "l", "", "Address and port to listen on")
    flag.StringVar(&gProxyServerSpec, "p", "", "Proxy servers to use, separated by commas. E.g. -p proxy1.tld.com:80,proxy2.tld.com:8080,proxy3.tld.com:80")
    flag.StringVar(&gDirects,         "d", "", "IP addresses to go direct")
    flag.StringVar(&gLogfile,         "f", "", "Log file")
    flag.StringVar(&gCpuProfile,      "c", "", "Write cpu profile to file")
    flag.StringVar(&gMemProfile,      "m", "", "Write mem profile to file")
    flag.IntVar(   &gVerbosity,       "v", 0,  "Control level of logging. v=1 results in debugging info printed to the log.\n")

    dirFuncs := buildDirectors(gDirects)
    director = getDirector(dirFuncs)
}

func versionString() (v string) {
    buildNum := strings.ToUpper(strconv.FormatInt(BUILDTIMESTAMP, 36))
    buildDate := time.Unix(BUILDTIMESTAMP, 0).Format(time.UnixDate)
    v = fmt.Sprintf("any_proxy %s (build %v, %v by %v@%v)", VERSION, buildNum, buildDate, BUILDUSER, BUILDHOST)
    return
}

func buildDirectors(gDirects string) ([]directorFunc) {
    // Generates a list of directorFuncs that are have "cached" values within
    // the scope of the functions.  

    dstrings := strings.Split(gDirects, ",")
    directors := make([]directorFunc, len(dstrings))

    for idx,dstrOrig := range dstrings {
        dstring := dstrOrig
        var dfunc directorFunc
        if strings.Contains(dstring, "/") {

            cidrIp, cidrIpNet, err := net.ParseCIDR(dstring)
            if err != nil {
                panic(fmt.Sprintf("\nUnable to parse CIDR string : %s : %s\n", dstring, err))
            }
            dfunc = func(ip string) bool {
                testIp := net.ParseIP(ip)
                if cidrIp.Equal(testIp.Mask(cidrIpNet.Mask)) {
                    return true
                }
                return false
            }
            directors[idx] = dfunc
        } else {
            dfunc = func(ip string) bool {
                if ip == dstring {
                    return true
                }
                return false
            }
            directors[idx] = dfunc
        }

    }
    return directors
}

func getDirector(directors []directorFunc) func(string) (bool, int) {
    // getDirector:
    // Returns a function(directorFunc) that loops through internal held 
    // directors evaluating each for possible matches.
    // 
    // directorFunc: 
    // Loops through directors and returns the (true, idx) where the index is 
    // the sequencial director that returned true. Else the function returns
    // (false, 0) if there are no directors to handle the ip.

    dFunc := func(ipStr string) (bool, int) {
        for idx,dfunc := range directors {
            if dfunc(ipStr) {
                return true, idx
            }
        }
        return false, 0
    }
    return dFunc
}

func setupProfiling() {
    // Make sure we have enough time to write profile's to disk, even if user presses Ctrl-C
    if gMemProfile == "" || gCpuProfile == "" {
        return
    }

    var profilef *os.File
    var err error
    if gMemProfile == "" {
        profilef, err = os.Create(gMemProfile)
        if err != nil {
            panic(err)
        }
    }

    if gCpuProfile != "" {
        f, err := os.Create(gCpuProfile)
        if err != nil {
            panic(err)
        }
        pprof.StartCPUProfile(f)
    }

    c := make(chan os.Signal, 1)
    signal.Notify(c, os.Interrupt)
    go func() {
        for _ = range c {
            if gCpuProfile != "" {
                pprof.StopCPUProfile()
            }
            if gMemProfile != "" {
                pprof.WriteHeapProfile(profilef)
                profilef.Close()
            }
            time.Sleep(5000 * time.Millisecond)
            os.Exit(0)
        }
    }()
}

func setupLogging() {
    if gLogfile == "" {
        gLogfile = DEFAULTLOG
    }

    log.SetLevel(log.INFO)
    if gVerbosity != 0 {
        log.SetLevel(log.DEBUG)
    }

    if err := log.OpenFile(gLogfile, log.FLOG_APPEND, 0644); err != nil {
        log.Fatalf("Unable to open log file : %s", err)
    }
}

func main() {
    flag.Parse()
    if gListenAddrPort == "" {
        flag.Usage()
        os.Exit(1)
    }

    runtime.GOMAXPROCS(runtime.NumCPU() / 2)
    setupLogging()
    setupProfiling()
    setupStats()

    dirFuncs := buildDirectors(gDirects)
    director = getDirector(dirFuncs)

    log.RedirectStreams()

    // if user gave us upstream proxies, check and see if they are alive
    if gProxyServerSpec != "" {
	checkProxies()
    }

    lnaddr, err := net.ResolveTCPAddr("tcp", gListenAddrPort)
    if err != nil {
        panic(err)
    }

    listener, err := net.ListenTCP("tcp", lnaddr)
    if err != nil {
        panic(err)
    }
    defer listener.Close()
    log.Infof("Listening for connections on %v\n", listener.Addr())

    for {
        conn, err := listener.AcceptTCP()
        if err != nil {
            log.Infof("Error accepting connection: %v\n", err)
            incrAcceptErrors()
            continue
        }
        incrAcceptSuccesses()
        go handleConnection(conn)
    }
}

func checkProxies() {
    gProxyServers = strings.Split(gProxyServerSpec, ",")
    // make sure proxies resolve and are listening on specified port
    for i, proxySpec := range gProxyServers {
        conn, err := dial(proxySpec)
        if err != nil {
            log.Infof("Test connection to %v: failed. Removing from proxy server list\n", proxySpec)
            a := gProxyServers[:i]
            b := gProxyServers[i+1:]
            gProxyServers = append(a, b...)
            continue
        }
        conn.Close()
        log.Infof("Added proxy server %v\n", proxySpec)
    }
    // do we have at least one proxy server?
    if len(gProxyServers) == 0 {
        msg := "None of the proxy servers specified are available. Exiting."
        log.Infof("%s\n", msg)
        fmt.Fprintf(os.Stderr, msg)
        os.Exit(1)
    }
}

func copy(dst io.ReadWriteCloser, src io.ReadWriteCloser, dstname string, srcname string) {
    if dst == nil {
        log.Debugf("copy(): oops, dst is nil!")
        return
    }
    if src == nil {
        log.Debugf("copy(): oops, src is nil!")
        return
    }
    _, err := io.Copy(dst, src)
    if err != nil {
        if operr, ok := err.(*net.OpError); ok {
            if srcname == "directserver" || srcname == "proxyserver" {
                log.Debugf("copy(): %s->%s: Op=%s, Net=%s, Addr=%v, Err=%v", srcname, dstname, operr.Op, operr.Net, operr.Addr, operr.Err)
            }
            if operr.Op == "read" {
                if srcname == "proxyserver" {
                    incrProxyServerReadErr()
                }
                if srcname == "directserver" {
                    incrDirectServerReadErr()
                }
            }
            if operr.Op == "write" {
                if srcname == "proxyserver" {
                    incrProxyServerWriteErr()
                }
                if srcname == "directserver" {
                    incrDirectServerWriteErr()
                }
            }
        }
    }
    dst.Close()
    src.Close()
}

func getOriginalDst(clientConn *net.TCPConn) (ipv4 string, port uint16, newTCPConn *net.TCPConn, err error) {
    if clientConn == nil {
        log.Debugf("copy(): oops, dst is nil!")
        err = errors.New("ERR: clientConn is nil")
        return
    }

    // test if the underlying fd is nil
    remoteAddr := clientConn.RemoteAddr()
    if remoteAddr == nil {
        log.Debugf("getOriginalDst(): oops, clientConn.fd is nil!")
        err = errors.New("ERR: clientConn.fd is nil")
        return
    }

    srcipport := fmt.Sprintf("%v", clientConn.RemoteAddr())

    newTCPConn = nil
    // net.TCPConn.File() will cause the receiver's (clientConn) socket to be placed in blocking mode.
    // The workaround is to take the File returned by .File(), do getsockopt() to get the original 
    // destination, then create a new *net.TCPConn by calling net.Conn.FileConn().  The new TCPConn
    // will be in non-blocking mode.  What a pain.
    clientConnFile, err := clientConn.File()
    if err != nil {
        log.Infof("GETORIGINALDST|%v->?->FAILEDTOBEDETERMINED|ERR: could not get a copy of the client connection's file object", srcipport)
        return
    } else {
        clientConn.Close()
    }

    // Get original destination
    // this is the only syscall in the Golang libs that I can find that returns 16 bytes
    // Example result: &{Multiaddr:[2 0 31 144 206 190 36 45 0 0 0 0 0 0 0 0] Interface:0}
    // port starts at the 3rd byte and is 2 bytes long (31 144 = port 8080)
    // IPv4 address starts at the 5th byte, 4 bytes long (206 190 36 45)
    addr, err :=  syscall.GetsockoptIPv6Mreq(int(clientConnFile.Fd()), syscall.IPPROTO_IP, SO_ORIGINAL_DST)
    log.Debugf("getOriginalDst(): SO_ORIGINAL_DST=%+v\n", addr)
    if err != nil {
        log.Infof("GETORIGINALDST|%v->?->FAILEDTOBEDETERMINED|ERR: getsocketopt(SO_ORIGINAL_DST) failed: %v", srcipport, err)
        return
    }
    newConn, err := net.FileConn(clientConnFile)
    if err != nil {
        log.Infof("GETORIGINALDST|%v->?->%v|ERR: could not create a FileConn fron clientConnFile=%+v: %v", srcipport, addr, clientConnFile, err)
        return
    }
    if _, ok := newConn.(*net.TCPConn); ok {
        newTCPConn = newConn.(*net.TCPConn)
        clientConnFile.Close()
    } else {
        errmsg := fmt.Sprintf("ERR: newConn is not a *net.TCPConn, instead it is: %T (%v)", newConn, newConn)
        log.Infof("GETORIGINALDST|%v->?->%v|%s", srcipport, addr, errmsg)
        err = errors.New(errmsg)
        return
    }

    ipv4 = itod(uint(addr.Multiaddr[4])) + "." + 
           itod(uint(addr.Multiaddr[5])) + "." + 
           itod(uint(addr.Multiaddr[6])) + "." + 
           itod(uint(addr.Multiaddr[7]))
    port = uint16(addr.Multiaddr[2]) << 8 + uint16(addr.Multiaddr[3])

    return
}

func dial(spec string) (*net.TCPConn, error) {
    host, port, err := net.SplitHostPort(spec)
    if err != nil {
        log.Infof("dial(): ERR: could not extract host and port from spec %v: %v", spec, err)
        return nil, err
    }
    remoteAddr, err := net.ResolveIPAddr("ip", host)
    if err != nil {
        log.Infof("dial(): ERR: could not resolve %v: %v", host, err)
        return nil, err
    }
    portInt, err := strconv.Atoi(port)
    if err != nil {
        log.Infof("dial(): ERR: could not convert network port from string \"%s\" to integer: %v", port, err)
        return nil, err
    }
    remoteAddrAndPort := &net.TCPAddr{IP: remoteAddr.IP, Port: portInt}
    var localAddr *net.TCPAddr
    localAddr = nil
    conn, err := net.DialTCP("tcp", localAddr, remoteAddrAndPort)
    if err != nil {
        log.Infof("dial(): ERR: could not connect to %v:%v: %v", remoteAddrAndPort.IP, remoteAddrAndPort.Port, err)
    }
    return conn, err
}

func handleDirectConnection(clientConn *net.TCPConn, ipv4 string, port uint16) {
    // TODO: remove
    log.Debugf("Enter handleDirectConnection: clientConn=%+v (%T)\n", clientConn, clientConn)

    if clientConn == nil {
        log.Debugf("handleDirectConnection(): oops, clientConn is nil!")
        return
    }

    // test if the underlying fd is nil
    remoteAddr := clientConn.RemoteAddr()
    if remoteAddr == nil {
        log.Debugf("handleDirectConnection(): oops, clientConn.fd is nil!")
        return
    }

    ipport := fmt.Sprintf("%s:%d", ipv4, port)
    directConn, err := dial(ipport)
    if err != nil {
        log.Infof("DIRECT|%v->%v|Could not connect, giving up: %v", clientConn.RemoteAddr(), directConn.RemoteAddr(), err)
        return
    }
    log.Debugf("DIRECT|%v->%v|Connected to remote end", clientConn.RemoteAddr(), directConn.RemoteAddr())
    incrDirectConnections()
    go copy(clientConn, directConn, "client", "directserver")
    go copy(directConn, clientConn, "directserver", "client")
}

func handleProxyConnection(clientConn *net.TCPConn, ipv4 string, port uint16) {
    var proxyConn net.Conn
    var err error
    var success bool = false
    var host string
    var headerXFF string = ""

    // TODO: remove
    log.Debugf("Enter handleProxyConnection: clientConn=%+v (%T)\n", clientConn, clientConn)

    if clientConn == nil {
        log.Debugf("handleProxyConnection(): oops, clientConn is nil!")
        return
    }

    // test if the underlying fd is nil
    remoteAddr := clientConn.RemoteAddr()
    if remoteAddr == nil {
        log.Debugf("handleProxyConnect(): oops, clientConn.fd is nil!")
        err = errors.New("ERR: clientConn.fd is nil")
        return
    }

    host, _, err = net.SplitHostPort(remoteAddr.String())
    if err == nil {
        headerXFF = fmt.Sprintf("X-Forwarded-For: %s\r\n", host)
    }

    for _, proxySpec := range gProxyServers {
        proxyConn, err = dial(proxySpec)
        if err != nil {
            log.Debugf("PROXY|%v->%v->%s:%d|Trying next proxy.", clientConn.RemoteAddr(), proxySpec, ipv4, port)
            continue
        }
        log.Debugf("PROXY|%v->%v->%s:%d|Connected to proxy\n", clientConn.RemoteAddr(), proxyConn.RemoteAddr(), ipv4, port)
        connectString := fmt.Sprintf("CONNECT %s:%d HTTP/1.0\r\n%s\r\n", ipv4, port, headerXFF)
        log.Debugf("PROXY|%v->%v->%s:%d|Sending to proxy: %s\n", clientConn.RemoteAddr(), proxyConn.RemoteAddr(), ipv4, port, strconv.Quote(connectString))
        fmt.Fprintf(proxyConn, connectString)
        status, err := bufio.NewReader(proxyConn).ReadString('\n')
        log.Debugf("PROXY|%v->%v->%s:%d|Received from proxy: %s", clientConn.RemoteAddr(), proxyConn.RemoteAddr(), ipv4, port, strconv.Quote(status))
        if err != nil {
            log.Infof("PROXY|%v->%v->%s:%d|ERR: Could not find response to CONNECT: err=%v. Trying next proxy", clientConn.RemoteAddr(), proxyConn.RemoteAddr(), ipv4, port, err)
            incrProxyNoConnectResponses()
            continue
        }
        if strings.Contains(status, "400") { // bad request
            log.Debugf("PROXY|%v->%v->%s:%d|Status from proxy=400 (Bad Request)", clientConn.RemoteAddr(), proxyConn.RemoteAddr(), ipv4, port)
            log.Debugf("%v: Response from proxy=400", proxySpec)
            incrProxy400Responses()
            copy(clientConn, proxyConn, "client", "proxyserver")
            return
        }
        if strings.Contains(status, "200") == false {
            log.Infof("PROXY|%v->%v->%s:%d|ERR: Proxy response to CONNECT was: %s. Trying next proxy.\n", clientConn.RemoteAddr(), proxyConn.RemoteAddr(), ipv4, port, strconv.Quote(status))
            incrProxyNon200Responses()
            continue
        }
        incrProxy200Responses()
        log.Debugf("PROXY|%v->%v->%s:%d|Proxied connection", clientConn.RemoteAddr(), proxyConn.RemoteAddr(), ipv4, port)
        success = true
        break
    }
    if proxyConn == nil {
        log.Debugf("handleProxyConnection(): oops, proxyConn is nil!")
        return
    }
    if success == false {
        log.Infof("PROXY|%v->UNAVAILABLE->%s:%d|ERR: Tried all proxies, but could not establish connection. Giving up.\n", clientConn.RemoteAddr(), ipv4, port)
        fmt.Fprintf(clientConn, "HTTP/1.0 503 Service Unavailable\r\nServer: go-https-proxy\r\nX-Go-Https-Proxy-Error: ERR_NO_PROXIES\r\n\r\n")
        clientConn.Close()
        return
    } 
    incrProxiedConnections()
    go copy(clientConn, proxyConn, "client", "proxyserver")
    go copy(proxyConn, clientConn, "proxyserver", "client")
}

func handleConnection(clientConn *net.TCPConn) {
    if clientConn == nil {
        log.Debugf("handleConnection(): oops, clientConn is nil")
        return
    }

    // test if the underlying fd is nil
    remoteAddr := clientConn.RemoteAddr()
    if remoteAddr == nil {
        log.Debugf("handleConnection(): oops, clientConn.fd is nil!")
        return
    }

    ipv4, port, clientConn, err := getOriginalDst(clientConn)
    if err != nil {
        log.Infof("handleConnection(): can not handle this connection, error occurred in getting original destination ip address/port: %+v\n", err)
        return
    }
    // If no upstream proxies were provided on the command line, assume all traffic should be sent directly
    if gProxyServerSpec == "" {
            handleDirectConnection(clientConn, ipv4, port)
            return
    } 
    // Evaluate for direct connection
    if ok,_ := director(ipv4); ok {
            handleDirectConnection(clientConn, ipv4, port)
            return
    }
    handleProxyConnection(clientConn, ipv4, port)
}

// from pkg/net/parse.go
// Convert i to decimal string.
func itod(i uint) string {
        if i == 0 {
                return "0"
        }

        // Assemble decimal in reverse order.
        var b [32]byte
        bp := len(b)
        for ; i > 0; i /= 10 {
                bp--
                b[bp] = byte(i%10) + '0'
        }

        return string(b[bp:])
}
